Add time wrapper to test agent delays in CI

joestringer opened the pull request:
Cilium's internal business logic relies on a highly parallel combination of reactive handlers for incoming information, "triggers" that ratelimit requests for processing to ensure Cilium does not over-consume resources, and "controllers" that periodically perform updates or resiliency checks of configured state. While in general most things are "eventually consistent", the presence of time-based triggers and controllers can introduce challenges when evaluating how Cilium will perform once the "eventual consistency" is resolved.

Taking an example of a situation where this eventual consistency caused issues that were not identified during pre-release testing, consider the issue fixed by https://github.com/cilium/cilium/pull/27327 . Cilium v1.14.0 was released with a bug where Cilium appeared to work correctly for the first several minutes following startup, then afterwards introduced connectivity disruption to specific peers. This was traced down to a logic delay tied to a 10 minute timer following startup which caused some state to be recomputed & configured in the system, ultimately causing the packet loss.

While better unit testing for the individual package could have caught this issue earlier, the issue was also triggered by the integration between the specific package and other logic in other packages. It is quite difficult to systematically identify time-based errors across the entire agent by relying purely on such testing in each package. This PR attempts to provide a more systematic safety net for timer-based issues by providing a hidden flag to override all timers within Cilium and ensuring that Cilium's CI runs with these timers set to a minimal value.

One of the interesting challenges with this PR is that it can be tempting for developers to rely on time-based tricks in order to ensure the execution order of specific pieces of logic. However, when the system is highly loaded, such mechanisms can become unreliable as an ordering enforcement mechanism. As a side-objective, this PR also hopes to make such tricks less viable on order to incentivize implementing better ordering mechanisms.

Review tips: The `treewide` commit has about 2/3 of the changes but is generated almost entirely from a script, so can be overlooked for initial review:
```
    treewide: Wrap time package with internal wrapper
...
178 files changed, 186 insertions(+), 191 deletions(-)
```

Related: https://github.com/cilium/cilium/issues/28844

joestringer committed:
pkg: Add time wrapper package

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Add flag for MaxInternalTimerDelay

This new flag will start to enforce maximum internal time durations and
timers for testing purposes. This relies on other commits that convert
internal packages over from upstream "time" to "pkg/time".

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Set externalTimerDelay to 5s in CI

Signed-off-by: Joe Stringer <joe@cilium.io>

sayboras commented on pkg/time/time.go:
```diff
@@ -0,0 +1,86 @@
+// aliases most underlying types, but allows overrides for testing purposes.
+//
+// Synced to go-1.20.7.
+package time
```
we can add the check similar to lock-check.sh if this new time package is supposed to be used in most of the cases.

joestringer commented on pkg/time/time.go:
```diff
@@ -0,0 +1,86 @@
+)
+
+var (
+	MaxInternalTimerDelay time.Duration
```
Make this link-time specified

joestringer replied:
Or maybe not, we don't want to accidentally break this for all CI images => can't get user input because the images diverge too far from prod images.

joestringer committed:
contrib: Add script to replace time with pkg/time

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
Makefile: Add check-time.sh to precheck

Enforce that time usage must use pkg/time going forwards so that CI can
detect eventual consistency issues related to timers.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
time: Support maximum Sleep periods

Implement wrapper functions for common sleep / timer functions in order
to place a maximum on the durations for these functions. Implemented via
a local variable in the new pkg/time package. This local variable is not
yet effective until future commits initialize it during agent startup.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
chore: Fix import ordering

I noticed a couple of instances where the Cilium imports were in the
wrong order. Fix them up.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
treewide: Wrap time package with internal wrapper

Generated with 'contrib/scripts/check-time.sh update'.

Earlier versions used:

    $ git grep -l "time\"" -- \
          ':!install/' ':!api/' ':!examples/' ':!cilium*' ':!clustermesh*' \
          ':!operator/' ':!bugtool' ':!tools' ':!test' ':!vendor' ':!plugins' \
          ':!hive' ':!*_test.go' ':!pkg/testutils' ':!pkg/time' ':!pkg/lock/' \
          ':!pkg/loadinfo' ':!pkg/health' ':!Documentation/' \
          ':!pkg/monitor' ':!pkg/k8s/client' ':!pkg/k8s/slim' \
      > files.txt
    $ cat files.txt \
      | xargs sed -i '/"time"/d; /"github.com\/cilium\/cilium\/.*"/a\
"github.com/cilium/cilium/pkg/time"'
    $ cat files.txt \
      | xargs dirname \
      | sort -u \
      | xargs go run golang.org/x/tools/cmd/goimports -w
    $ cat files.txt \
      | xargs git add

Signed-off-by: Joe Stringer <joe@cilium.io>

danehans commented on pkg/time/time.go:
```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+// package time is a wrapper for the upstream system library for time.Time that
```
nit: space between `.` and `Time`

joestringer replied:
Actually this was originally intentional as `time.Time` but the lib ended up wrapping more than just `time.Time` so I'll fix it up to refer that it just wraps the whole upstream library.

mhofstetter commented on contrib/scripts/check-time.sh:
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env bash
```
I wonder whether we could replace this checkfile with an existing linter provided by golangci-lint.

We're already using `gomodguard` to prevent the usage of some go modules -> https://github.com/cilium/cilium/blob/main/.golangci.yaml#L66-L81.

[DepGuard](https://golangci-lint.run/usage/linters/#depguard) should be able to prevent access to std library packages and provides the possibility to define regexes for the directories.

Linter: https://github.com/OpenPeeDeeP/depguard

joestringer replied:
Do any of those options provide a way to script updating the dependency?

joestringer replied:
While writing this script, I did feel like it would be better to have a more generic mechanism for this sort of linting / updating, since there are various scripts that work in a similar way and they really just look for some pattern and reject it, and maybe there's a way to update. But the update is actually the most complicated case.

There was also a point where I thought that it was a mistake to write this linter/updater in bash instead of Go :sweat_smile: At this point I think this is functional enough that I'd like to move on. The incremental benefit of reworking this part of the solution seems minimal to me.

Other than that, while I'm _open_ to more golangci-lint, on a personal level I'm not a fan because of #25155. Taking 8-10 minutes to complain about something a good `grep` can do in less than a second is painful.

danehans replied:
+1 on proceeding since the script is functional but it might be a good idea to capture the Go/gomodguard refactor in an issue.

mhofstetter replied:
> Do any of those options provide a way to script updating the dependency?

No it would only cover the linting (check) part

> At this point I think this is functional enough that I'd like to move on.

Totally fine for me :+1: Just thought it's worth to drop a comment

joestringer replied:
Yep good feedback and worth considering. Thanks for raising this discussion item.

gandro commented on test/helpers/kubectl.go:
```diff
@@ -122,6 +122,8 @@ var (
 
 		"ipam.operator.clusterPoolIPv6PodCIDRList": "fd02::/112",
+
+		"extraConfig.max-internal-timer-delay": "5s",
```
Should we also add this to the GitHub Actions default values here:
https://github.com/cilium/cilium/blob/main/.github/actions/helm-default/action.yaml

joestringer replied:
I think I'll file an issue with this and other followups if that's alright with you. There's a few more small iterative items we can do on top like:
- Apply this change to more than just cilium-agent (operator etc.)
- See if there are additional packages that can reuse the time wrapper.
- Pick a different value than 5s

The PR is currently pulling in a lot of owners so I wanted to make sure we get together an implementation that works and transitions the codebase over, then we can iterate and tweak the details.

joestringer replied:
Filed https://github.com/cilium/cilium/issues/28844

tommyp1ckles commented on pkg/time/time.go:
```diff
@@ -72,3 +66,58 @@ type (
+// option.MaxInternalTimerDelay.
+func After(d Duration) <-chan Time {
+	if MaxInternalTimerDelay > 0 && d > MaxInternalTimerDelay {
+		d = MaxInternalTimerDelay
```
Will this only be used in testing, should we emit a log that something is having time.After/Sleep capped?

joestringer replied:
No harm emitting a debug log here yeah. Could help folks understand better in future if they are unaware of this change.

joestringer replied:
Added to https://github.com/cilium/cilium/issues/28844

derailed reviewed (approved):
@joestringer Nice work! Just wish there was a better way to handle this ;(

tommyp1ckles commented on contrib/scripts/check-time.sh:
```diff
@@ -0,0 +1,110 @@
+  "rand"
+  "types"
+
+  # May need subsequent evaluation to detect resiliency issues in CI
```
Is this a todo?

joestringer replied:
More or less yeah. The PR provides a good basis for now, there's some packages I stopped short of since they may have more nuanced interactions. I figured that for those, it'd be better to iterate on top once the initial bigger PR is agreed upon.

Added to the tasklist in https://github.com/cilium/cilium/issues/28844 .
